### PR TITLE
feat: `info tools` now returns a lot more information

### LINKED ISSUE #833: `jbang info tools` should expose more info
Would be nice if `jbang info tools`  could return
- The Java version
- potential path to JDK to use
- Java options to apply
- associated sources (//SOURCES)
- associated files (//FILES)

Ideally, the position where an error is detected (eg. bad dependency).

See https://github.com/fbricon/jbang.eclipse/blob/master/jbang.eclipse.core/src/main/java/io/jbang/eclipse/core/internal/process/JBangExecution.java#L90 for context

### PR DESCRIPTION
**Pull request by quintesse**
Fixes #833

Eg:

```json
$ jbang info tools itests/resource.java
{
  "originalResource": "itests/resource.java",
  "backingResource": "/home/tschotan/projects/jbang/jbang/itests/resource.java",
  "applicationJar": "/home/tschotan/.jbang/cache/jars/resource.java.dd614e274e05a4b3e200963f43ba65e8ee416d7a0955957a1cc93f6f7c7cd224.jar",
  "resolvedDependencies": [],
  "availableJdkPath": "/home/tschotan/.sdkman/candidates/java/current",
  "files": [
    {
      "originalResource": "itests/resource.properties",
      "backingResource": "/home/tschotan/projects/jbang/jbang/itests/resource.properties"
    },
    {
      "originalResource": "itests/resource.properties",
      "backingResource": "/home/tschotan/projects/jbang/jbang/itests/resource.properties",
      "target": "renamed.properties"
    },
    {
      "originalResource": "itests/resource.properties",
      "backingResource": "/home/tschotan/projects/jbang/jbang/itests/resource.properties",
      "target": "META-INF/application.properties"
    }
  ]
}
```

```json
$ jbang info tools itests/one.java     
{
  "originalResource": "itests/one.java",
  "backingResource": "/home/tschotan/projects/jbang/jbang/itests/one.java",
  "applicationJar": "/home/tschotan/.jbang/cache/jars/one.java.a4c9946d3aa4a238c7c44b87c38817ab8aa6d5e48a65cd6ad8cf23b6cf5dade0.jar",
  "resolvedDependencies": [
    "/home/tschotan/.m2/repository/info/picocli/picocli/4.5.0/picocli-4.5.0.jar",
    "/home/tschotan/.m2/repository/org/kohsuke/github-api/1.116/github-api-1.116.jar",
    "/home/tschotan/.m2/repository/org/apache/commons/commons-lang3/3.9/commons-lang3-3.9.jar",
    "/home/tschotan/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.10.2/jackson-databind-2.10.2.jar",
    "/home/tschotan/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.10.2/jackson-annotations-2.10.2.jar",
    "/home/tschotan/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.10.2/jackson-core-2.10.2.jar",
    "/home/tschotan/.m2/repository/commons-io/commons-io/2.4/commons-io-2.4.jar"
  ],
  "availableJdkPath": "/home/tschotan/.sdkman/candidates/java/current",
  "sources": [
    {
      "originalResource": "itests/Two.java",
      "backingResource": "/home/tschotan/projects/jbang/jbang/itests/Two.java",
      "sources": [
        {
          "originalResource": "itests/gh_fetch_release_assets.java",
          "backingResource": "/home/tschotan/projects/jbang/jbang/itests/gh_fetch_release_assets.java",
          "dependencies": [
            "info.picocli:picocli:4.5.0",
            "org.kohsuke:github-api:1.116"
          ]
        },
        {
          "originalResource": "itests/gh_release_stats.java",
          "backingResource": "/home/tschotan/projects/jbang/jbang/itests/gh_release_stats.java",
          "dependencies": [
            "info.picocli:picocli:4.5.0",
            "org.kohsuke:github-api:1.116"
          ]
        }
      ]
    }
  ]
}
```

### CODE REVIEW

File: nohup.out [hunk 1]
```diff
@@ -0,0 +1,40 @@
+2021-05-03 22:50:49,704 [   2507]   WARN - ntellij.ui.svg.SvgCacheManager - Icon Cache will be recreated or previous version of data reused (db=/home/tschotan/.cache/JetBrains/IdeaIC2021.1/icons-v2.db)
```
Comment (fbricon) [inline nohup.out:1]: huh? accidentally added?

Comment (quintesse) [reply nohup.out:1]: Oops , thanks :-)

**Comment by fbricon** [comment]
so you can have a jbang file that declares its dependencies and other files that have a different set of dependencies? Then what's the final classpath?

**Comment by quintesse** [comment]
> Then what's the final classpath?

@fbricon the `resolvedDependencies` is the final classpath and will only appear in the toplevel object. The `dependencies` are the (unresolved) dependencies that each file declares.

**Comment by fbricon** [comment]
What happens in the case of circular dependencies? one.java[source=two.java] + two.java[source=one.java]

**Comment by quintesse** [comment]
> What happens in the case of circular dependencies?

It breaks :-)

File: src/main/java/dev/jbang/cli/Info.java [hunk 2]
```diff
@@ -33,33 +61,77 @@
...
+				files = refs.stream()
+							.map(ref -> new ResourceFile(ref))
+							.collect(Collectors.toList());
+			}
+			List<ScriptSource> srcs = ss.collectSources();
+			if (!srcs.isEmpty()) {
+				sources = srcs	.stream()
+								.map(s -> new ScriptInfo(s, null))
+								.collect(Collectors.toList());
+			}
```
Comment (fbricon) [inline src/main/java/dev/jbang/cli/Info.java:110]: interesting indentation style ;-)

**Comment by fbricon** [comment]
How do compile options look like?

**Comment by fbricon** [comment]
orthogonal issue: would be nice if the CI build generated an archive for Mac too https://github.com/jbangdev/jbang/actions/runs/820120167

File: src/main/java/dev/jbang/cli/Info.java (hunk 2 above)
Comment (quintesse) [reply src/main/java/dev/jbang/cli/Info.java:110]: That's @maxandersen enforcing his style with Spotless. I hate it myself but I squint my eyes when looking at the code and try not to see it :-D

**Comment by quintesse** [comment]
> How do compile options look like?

@fbricon this is what runtime options look like, there's a `compileOptions` that work the same:

```json
  "runtimeOptions": [
    "-Dfoo\u003dbar",
    "-Dbar\u003daap noot mies"
  ]
```

File: src/main/java/dev/jbang/cli/Info.java (hunk 2 above)
Comment (fbricon) [reply src/main/java/dev/jbang/cli/Info.java]: that explains it. Wikings are weird ;-)

**Comment by fbricon** [comment]
>     "-Dfoo\u003dbar",
>     "-Dbar\u003daap noot mies"

Is the url encoding really necessary?

**Comment by maxandersen** [comment]
looks good - I'll merge it but we might consider including native image result too if available.

**Pull request merged**